worlds/ror2/Rules.py

```diff
@@ -7,27 +7,37 @@ class RiskOfRainLogic(LogicMixin):
     def _ror_has_items(self, player: int, amount: int) -> bool:
         count: int = self.item_count("Common Item", player) + self.item_count("Uncommon Item", player) + \
                      self.item_count("Legendary Item", player) + self.item_count("Boss Item", player) + \
-                     self.item_count("Lunar Item", player) + self.item_count("Equipment", player)
+                     self.item_count("Lunar Item", player) + self.item_count("Equipment", player) + \
+                     self.item_count("Dio's Best Friend", player)
         return count >= amount
 
 
 def set_rules(world: MultiWorld, player: int):
-    total_checks = world.total_locations[player]
     # divide by 5 since 5 levels (then commencement)
-    items_per_level = total_checks / 5
-    leftover = total_checks % 5
+    items_per_level = max(int(world.total_locations[player] / 5 / (world.item_pickup_step[player]+1)), 1)
 
-    set_rule(world.get_location("Level One", player),
-             lambda state: state._ror_has_items(player, items_per_level + leftover))
+    # lock item pickup access based on level completion
+    for i in range(1, items_per_level):
+        set_rule(world.get_location(f"Item Pickup {i}", player), lambda state: True)
+    for i in range(items_per_level, 2*items_per_level):
+        set_rule(world.get_location(f"Item Pickup {i}", player), lambda state: state.has("Beat Level One", player))
+    for i in range(2*items_per_level, 3*items_per_level):
+        set_rule(world.get_location(f"Item Pickup {i}", player), lambda state: state.has("Beat Level Two", player))
+    for i in range(3*items_per_level, 4*items_per_level):
+        set_rule(world.get_location(f"Item Pickup {i}", player), lambda state: state.has("Beat Level Three", player))
+    for i in range(4*items_per_level, world.total_locations[player]+1):
+        set_rule(world.get_location(f"Item Pickup {i}", player), lambda state: state.has("Beat Level Four", player))
+
+    # require items to beat each stage
     set_rule(world.get_location("Level Two", player),
-             lambda state: state._ror_has_items(player, items_per_level) and state.has("Beat Level One", player))
+             lambda state: state.has("Beat Level One", player) and state._ror_has_items(player, items_per_level))
     set_rule(world.get_location("Level Three", player),
-             lambda state: state._ror_has_items(player, items_per_level) and state.has("Beat Level Two", player))
+             lambda state: state._ror_has_items(player, 2 * items_per_level) and state.has("Beat Level Two", player))
```

Comment:
If `state.has("Beat Level Two")` then they already have at least all the items from level one. Do you need to specify `x * items_per_level`?

Reply:
Yes. This ensures the RoR2 player is able to receive those since they're deemed required for progression.

Reply:
Yeah, I get it now.